File: 2026-bogota-agent-sdk/typescript/04_mcp.ts
// 04 · Un servidor MCP entero. gmail_mcp.ts publica dos tools: buscar y leer.
// Se enchufa igual que una tool suelta: una entrada más en mcpServers.
import "./entorno.ts"; // carga ANTHROPIC_API_KEY de .env
import { query } from "@anthropic-ai/claude-agent-sdk";
import { servidorGmail } from "./gmail_mcp.ts";

const conversacion = query({
  prompt:
    "Hoy es sábado 19 de septiembre de 2026. Busca en mi correo los mensajes " +
    "de aerolíneas desde el 12 de septiembre y dime qué encontraste.",
  options: {
    model: "claude-sonnet-5",
    systemPrompt:
      "Eres un asistente personal. Respondes en español, corto y con datos. " +
      "Usa tus tools y skills; no adivines fechas ni reservas.",
    tools: [],
    mcpServers: { gmail: servidorGmail },

    // La clave de mcpServers es el nombre que aparece en el id de cada tool.
    allowedTools: ["mcp__gmail__buscar", "mcp__gmail__leer"],

    settingSources: [],
    strictMcpConfig: true, // sin esto entran los conectores de claude.ai y el agente iría a tu Gmail real
    maxTurns: 8,
    maxBudgetUsd: 0.10,
    env: { ...process.env, CLAUDE_CODE_DISABLE_AUTO_MEMORY: "1" },
  },
});

for await (const m of conversacion) {
  if (m.type === "system" && m.subtype === "init") {
    const estado = m.mcp_servers.map((s) => `${s.name}=${s.status}`).join(", ");
    console.log(`[init] servidores MCP: ${estado}`);
  }
  if (m.type === "assistant") {
    for (const b of m.message.content) {
      if (b.type === "tool_use") console.log(`-> ${b.name} ${JSON.stringify(b.input)}`);
      if (b.type === "text" && b.text.trim()) console.log(`\nAgente: ${b.text.trim()}`);
    }
  }
  if (m.type === "result" && m.subtype === "success") {
    console.log(`\nCosto: $${m.total_cost_usd.toFixed(4)}`);
  }
}

File: 2026-bogota-agent-sdk/.claude/skills/crear-agente-sdk/references/ejemplos/python/04_mcp.py
"""04 · MCP — tools que no escribiste tú, en un proceso aparte.

`gmail_mcp.py` es un servidor MCP falso con la bandeja de correo: buscar y leer.
El SDK lo lanza como subproceso y habla con él por stdio. Para el agente no hay
diferencia entre esas tools y `hoy()`: las ve igual, con nombre mcp__gmail__*.

Dos cosas que salvan la demo en una sala:
  strict_mcp_config=True   sin esto entran los conectores de tu cuenta claude.ai
                           (Gmail real, Drive, Calendar) y el agente los usa
  mcp_servers inline       porque strict_mcp_config también ignora el .mcp.json
                           de la carpeta: si quieres un servidor, decláralo aquí

Mira en el ARRANQUE el estado de cada servidor: connected · pending · needs-auth.

    uv run python 04_mcp.py
"""

import asyncio
import sys
from pathlib import Path

from claude_agent_sdk import ClaudeAgentOptions, create_sdk_mcp_server, query, tool

from mostrar import linea

AQUI = Path(__file__).parent.resolve()
PREGUNTA = "Revisa mis correos de la última semana y dime cuándo es mi próximo vuelo."

SYSTEM_PROMPT = """Eres un asistente personal. Respondes en español, corto y con datos.
Usa tus tools y skills; no adivines fechas ni reservas."""


@tool("hoy", "La fecha de hoy. Pídela siempre antes de razonar sobre fechas.", {})
async def hoy(args):
    return {"content": [{"type": "text", "text": "sábado 19 de septiembre de 2026"}]}


RELOJ = create_sdk_mcp_server(name="reloj", version="1.0.0", tools=[hoy])


async def main():
    opciones = ClaudeAgentOptions(
        model="claude-sonnet-5",
        system_prompt=SYSTEM_PROMPT,
        tools=[],
        mcp_servers={
            "reloj": RELOJ,                                             # dentro de este proceso
            "gmail": {"command": sys.executable,                         # proceso aparte, por stdio
                      "args": [str(AQUI / "gmail_mcp.py")]},
        },
        allowed_tools=["mcp__reloj__hoy", "mcp__gmail__buscar", "mcp__gmail__leer"],
        setting_sources=[],
        strict_mcp_config=True,
        env={"CLAUDE_CODE_DISABLE_AUTO_MEMORY": "1"},
        max_turns=10,
        max_budget_usd=0.20,
        stderr=lambda _: None,
    )

    print(f"PREGUNTA  {PREGUNTA}\n")
    async for m in query(prompt=PREGUNTA, options=opciones):
        linea(m)


asyncio.run(main())
